File: src/main/scala/com/exasol/cloudetl/helper/ParquetSchemaConverter.scala
package com.exasol.cloudetl.helper

import java.util.Locale.ENGLISH

import com.exasol.cloudetl.data.ExaColumnInfo

import org.apache.parquet.schema.LogicalTypeAnnotation._
import org.apache.parquet.schema.MessageType
import org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName
import org.apache.parquet.schema.Type
import org.apache.parquet.schema.Type.Repetition
import org.apache.parquet.schema.Types

/**
 * A class that converts Exasol column names to Parquet schema.
 *
 * @param isLowercaseSchemaEnabled a boolean flag to make Parquet schema fields lowercase
 */
final case class ParquetSchemaConverter(isLowercaseSchemaEnabled: Boolean) extends JavaClassTypes {
  import ParquetSchemaConverter._

  /**
   * Given the Exasol column information returns Parquet [[org.apache.parquet.schema.MessageType]].
   *
   * @param columns a list of Exasol column information
   * @param schemaName a Parquet schema name
   * @return a Parquet message type
   */
  def createParquetMessageType(columns: Seq[ExaColumnInfo], schemaName: String): MessageType = {
    val types = columns.map(columnToParquetField(_))
    new MessageType(schemaName, types: _*)
  }

  private[this] def columnToParquetField(columnInfo: ExaColumnInfo): Type = {
    val columnName = convertIdentifier(columnInfo.name)
    getParquetType(columnName, columnInfo)
  }

  private[this] def convertIdentifier(columnName: String): String = {
    val convertedName = columnName.replace("\"", "")
    if (isLowercaseSchemaEnabled) {
      convertedName.toLowerCase(ENGLISH)
    } else {
      convertedName
    }
  }

  private[this] def getParquetType(columnName: String, columnInfo: ExaColumnInfo): Type = {
    val columnType = columnInfo.`type`
    val repetition = if (columnInfo.isNullable) Repetition.OPTIONAL else Repetition.REQUIRED
    columnType match {
      case `jInteger` =>
        if (columnInfo.precision == 0) {
          Types
            .primitive(PrimitiveTypeName.INT32, repetition)
            .named(columnName)
        } else {
          require(
            columnInfo.precision <= DECIMAL_MAX_INT_DIGITS,
            s"Got an 'Integer' type with more than '$DECIMAL_MAX_INT_DIGITS' precision."
          )
          Types
            .primitive(PrimitiveTypeName.INT32, repetition)
            .as(decimalType(columnInfo.scale, columnInfo.precision))
            .named(columnName)
        }
      case `jLong` =>
        if (columnInfo.precision == 0) {
          Types
            .primitive(PrimitiveTypeName.INT64, repetition)
            .named(columnName)
        } else {
          require(
            columnInfo.precision <= DECIMAL_MAX_LONG_DIGITS,
            s"Got a 'Long' type with more than '$DECIMAL_MAX_LONG_DIGITS' precision."
          )
          Types
            .primitive(PrimitiveTypeName.INT64, repetition)
            .as(decimalType(columnInfo.scale, columnInfo.precision))
            .named(columnName)
        }
      case `jBigDecimal` =>
        Types
          .primitive(PrimitiveTypeName.FIXED_LEN_BYTE_ARRAY, repetition)
          .length(PRECISION_TO_BYTE_SIZE(columnInfo.precision - 1))
          .as(decimalType(columnInfo.scale, columnInfo.precision))
          .named(columnName)
      case `jDouble` =>
        Types
          .primitive(PrimitiveTypeName.DOUBLE, repetition)
          .named(columnName)
      case `jString` =>
        if (columnInfo.length > 0) {
          Types
            .primitive(PrimitiveTypeName.BINARY, repetition)
            .length(columnInfo.length)
            .as(stringType())
            .named(columnName)
        } else {
          Types
            .primitive(PrimitiveTypeName.BINARY, repetition)
            .as(stringType())
            .named(columnName)
        }
      case `jBoolean` =>
        Types
          .primitive(PrimitiveTypeName.BOOLEAN, repetition)
          .named(columnName)

      case `jSqlDate` =>
        Types
          .primitive(PrimitiveTypeName.INT32, repetition)
          .as(dateType())
          .named(columnName)

      case `jSqlTimestamp` =>
        Types
          .primitive(PrimitiveTypeName.INT96, repetition)
          .named(columnName)
      case _ =>
        throw new IllegalArgumentException(s"Cannot convert Exasol type '$columnType' to Parquet type.")
    }
  }

}

object ParquetSchemaConverter {
  val DECIMAL_MAX_PRECISION: Int = 38
  val DECIMAL_MAX_INT_DIGITS: Int = 9
  val DECIMAL_MAX_LONG_DIGITS: Int = 18

  // Maps the precision value into the number of bytes.
  // Adapted from:
  //  - org.apache.hadoop.hive.ql.io.parquet.serde.ParquetHiveSerDe.java
  val PRECISION_TO_BYTE_SIZE: Seq[Int] = {
    for {
      prec <- 1 to 38 // [1 .. 38]
      power = Math.pow(10, prec.toDouble) // scalastyle:ignore magic.number
      size = Math.ceil((Math.log(power - 1) / Math.log(2) + 1) / 8)
    } yield size.toInt
  }

}

File: src/main/java/com/exasol/cloudetl/helper/ParquetSchemaConverter.java
package com.exasol.cloudetl.helper;

import static java.util.Locale.ENGLISH;
import static org.apache.parquet.schema.LogicalTypeAnnotation.*;
import static org.apache.parquet.schema.PrimitiveType.PrimitiveTypeName.*;

import java.util.*;

import org.apache.parquet.schema.*;

import com.exasol.cloudetl.ScalaConverters;
import com.exasol.cloudetl.data.ExaColumnInfo;
import com.exasol.errorreporting.ExaError;

/** Converts Exasol column metadata to a Parquet schema. */
public final class ParquetSchemaConverter {
    /** Decimal maximum precision. */
    public static final int DECIMAL_MAX_PRECISION = 38;
    /** Maximum decimal precision for int32. */
    public static final int DECIMAL_MAX_INT_DIGITS = 9;
    /** Maximum decimal precision for int64. */
    public static final int DECIMAL_MAX_LONG_DIGITS = 18;
    /** Precision-to-byte-size mapping. */
    public static final List<Integer> PRECISION_TO_BYTE_SIZE = createPrecisionToByteSize();

    private final boolean lowercaseSchemaEnabled;

    /** Create a converter. */
    public ParquetSchemaConverter(final boolean lowercaseSchemaEnabled) {
        this.lowercaseSchemaEnabled = lowercaseSchemaEnabled;
    }

    /** Factory for Java callers. */
    public static ParquetSchemaConverter apply(final boolean lowercaseSchemaEnabled) {
        return new ParquetSchemaConverter(lowercaseSchemaEnabled);
    }

    /** Create a Parquet message type. */
    public MessageType createParquetMessageType(final scala.collection.immutable.Seq<ExaColumnInfo> columns,
            final String schemaName) {
        return createParquetMessageType(ScalaConverters.asJavaList(columns), schemaName);
    }

    /** Create a Parquet message type. */
    public MessageType createParquetMessageType(final List<ExaColumnInfo> columns, final String schemaName) {
        final List<Type> types = new ArrayList<>(columns.size());
        for (final ExaColumnInfo column : columns) {
            types.add(columnToParquetField(column));
        }
        return new MessageType(schemaName, types);
    }

    private Type columnToParquetField(final ExaColumnInfo columnInfo) {
        return getParquetType(convertIdentifier(columnInfo.name), columnInfo);
    }

    private String convertIdentifier(final String columnName) {
        final String convertedName = columnName.replace("\"", "");
        return this.lowercaseSchemaEnabled ? convertedName.toLowerCase(ENGLISH) : convertedName;
    }

    private Type getParquetType(final String columnName, final ExaColumnInfo columnInfo) {
        final Class<?> columnType = columnInfo.type;
        final Type.Repetition repetition = columnInfo.isNullable ? Type.Repetition.OPTIONAL : Type.Repetition.REQUIRED;
        if (Objects.equals(columnType, JavaClassTypes.J_INTEGER)) {
            return getIntegerType(columnName, columnInfo, repetition);
        } else if (Objects.equals(columnType, JavaClassTypes.J_LONG)) {
            return getLongType(columnName, columnInfo, repetition);
        } else if (Objects.equals(columnType, JavaClassTypes.J_BIG_DECIMAL)) {
            return getPrimitiveType(columnName, FIXED_LEN_BYTE_ARRAY, repetition,
                    PRECISION_TO_BYTE_SIZE.get(columnInfo.precision - 1), decimalType(columnInfo.scale, columnInfo.precision));
        } else if (Objects.equals(columnType, JavaClassTypes.J_DOUBLE)) {
            return getPrimitiveType(columnName, DOUBLE, repetition, null, null);
        } else if (Objects.equals(columnType, JavaClassTypes.J_STRING)) {
            return getPrimitiveType(columnName, BINARY, repetition, columnInfo.length > 0 ? columnInfo.length : null,
                    stringType());
        } else if (Objects.equals(columnType, JavaClassTypes.J_BOOLEAN)) {
            return getPrimitiveType(columnName, BOOLEAN, repetition, null, null);
        } else if (Objects.equals(columnType, JavaClassTypes.J_SQL_DATE)) {
            return getPrimitiveType(columnName, INT32, repetition, null, dateType());
        } else if (Objects.equals(columnType, JavaClassTypes.J_SQL_TIMESTAMP)) {
            return getPrimitiveType(columnName, INT96, repetition, null, null);
        }
        throw new IllegalArgumentException(ExaError.messageBuilder("F-CSE-22")
                .message("Cannot convert Exasol type {{TYPE}} to Parquet type.").parameter("TYPE", String.valueOf(columnType))
                .ticketMitigation().toString());
    }

    private Type getIntegerType(final String columnName, final ExaColumnInfo columnInfo,
            final Type.Repetition repetition) {
        if (columnInfo.precision == 0) {
            return getPrimitiveType(columnName, INT32, repetition, null, null);
        }
        require(columnInfo.precision <= DECIMAL_MAX_INT_DIGITS,
                "Got an 'Integer' type with more than '" + DECIMAL_MAX_INT_DIGITS + "' precision.");
        return getPrimitiveType(columnName, INT32, repetition, null, decimalType(columnInfo.scale, columnInfo.precision));
    }

    private Type getLongType(final String columnName, final ExaColumnInfo columnInfo, final Type.Repetition repetition) {
        if (columnInfo.precision == 0) {
            return getPrimitiveType(columnName, INT64, repetition, null, null);
        }
        require(columnInfo.precision <= DECIMAL_MAX_LONG_DIGITS,
                "Got a 'Long' type with more than '" + DECIMAL_MAX_LONG_DIGITS + "' precision.");
        return getPrimitiveType(columnName, INT64, repetition, null, decimalType(columnInfo.scale, columnInfo.precision));
    }

    private Type getPrimitiveType(final String name, final PrimitiveType.PrimitiveTypeName primitiveType,
            final Type.Repetition repetition, final Integer length, final LogicalTypeAnnotation logicalType) {
        Types.PrimitiveBuilder<PrimitiveType> resultType = Types.primitive(primitiveType, repetition);
        if (length != null) {
            resultType = resultType.length(length);
        }
        if (logicalType != null) {
            resultType = resultType.as(logicalType);
        }
        return resultType.named(name);
    }

    private static List<Integer> createPrecisionToByteSize() {
        final List<Integer> result = new ArrayList<>(DECIMAL_MAX_PRECISION);
        for (int precision = 1; precision <= DECIMAL_MAX_PRECISION; precision++) {
            final double power = Math.pow(10, precision);
            final double size = Math.ceil((Math.log(power - 1) / Math.log(2) + 1) / 8);
            result.add((int) size);
        }
        return List.copyOf(result);
    }

    private static void require(final boolean condition, final String message) {
        if (!condition) {
            throw new IllegalArgumentException("requirement failed: " + message);
        }
    }
}
